Fix job duplication

RaptDept opened the pull request:
Changes:
- Allow ack before calling handler to prevent long-running jobs from being duplicated
- Delay payload parsing until necessary
- Fix: error-retry loop ignores handler return values

Notes:
- I made the new behavior opt-in instead of completely replacing the current one, so that it doesn't break existing clients. It's disabled by default.
- I'm not sure what to name the new option (`workerAckOnStart`/`ackOnStart`), if there are better ideas I'd be happy to change it.
- I'm not sure if it should also be in the `queueConfig` parameter or have its own parameter.

RaptDept committed:
Delay payload parsing until necessary

RaptDept committed:
Allow ack before calling handler to prevent long-running jobs from being duplicated

maktouch commented:
It shouldn't be on the workers level - this should be on the handler's level. Just like how you can set the retries on the file itself. 

```js
module.exports = {
  ackOnStart: true,
  work: async function(payload) {
    //
  },
}
```

maktouch commented:
and we should maybe do something like a default key, where you can set the default retries, ackOnStart, etc.

RaptDept commented:
It looks like handler-level retries are already checked on the handler object itself: `const retries = getRetries(message.attributes.retries, handler.retries);`

Should I copy this behavior for `delayed` and `ackOnStart`, or should I move all three to something like `handler.default = {}`?

maktouch commented:
Please copy this behaviour for `delayed` and `ackOnStart`

RaptDept committed:
- Remove worker-level `ackOnStart`
- Copy `retries` behavior for `delayed` and `ackOnStart`
-- Add handler-level `delayed`
-- Add handler-level and message-level `ackOnStart`

RaptDept commented:
Adding a default handler `delayed` means I had to copy the `delayed` object parsing behavior from the Publisher class into the Worker class.

I think I need to make one more change to make sure `put`/`retry` treats a handler-level `delayed` the same way it treats a message-level `delayed`.

RaptDept committed:
Keep absolute `delayed` value the same on republish

maktouch reviewed (approved):
Looks good to me. Does it break existing stuff?

RaptDept commented:
> Looks good to me. Does it break existing stuff?

Nope, no need to change anything on existing clients
